core/psh/psh.c

```diff
@@ -972,7 +975,113 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_logincredentget(char *buff, int ispasswd, int maxinlen)
+{
+	int inputlen = -1;
+	char c;
+
+	for (;;) {
+		if (inputlen < 0) {
+			if (ispasswd)
+				fprintf(stdout, "Password: ");
+			else
+				fprintf(stdout, "Login: ");
+
+			fflush(stdout);
+			inputlen++;
+		}
+
+		if (read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+			
+		if (c == '\n') {
+			if (inputlen <= maxinlen && inputlen > 0) {
+				buff[inputlen] = '\0';
+				return --inputlen;
+			}
+			else {
+				inputlen = -1;
+				if (ispasswd) {
+					memset(buff, '\0', sizeof(char)*maxinlen);
+					return 0;
+				}
+				else
+					continue;
+			}
+		}
+		else if (isprint(c)) {
+			if (inputlen < maxinlen)
+				buff[inputlen] = c;
+			inputlen++;
+			continue;
+		}
+	}
+	return -EINTR;
+}
+
+
+static int psh_login(void)
+{
+	int err = 0, auth = -EACCES;
+	const int maxlen = 32;
+	char username[maxlen+1], passwd[maxlen+1], *shadow;
+	struct passwd *userdata;
+	struct termios origattr, ttyattr;
+
+	/* get/prepare tty settings */
+	if ((err = tcgetattr(STDIN_FILENO, &origattr)) < 0 || (err = tcgetattr(STDIN_FILENO, &ttyattr)) < 0) {
+		sleep(1);
+		fprintf(stderr, "psh: failed to save terminal settings\n");
+		return -ENOTTY;
+	}
+	if ((err = isatty(STDOUT_FILENO)) == 0) {
+		sleep(1);
+		fprintf(stderr, "psh: unable to login, not a tty\n");
+		return -ENOTTY;
+	}
+	ttyattr.c_lflag |= (ICANON | ECHO);
+	tcsetattr(STDIN_FILENO, TCSAFLUSH, &ttyattr);
+
+	/* Get login from user */
+	psh_logincredentget(username, 0, maxlen);
+	userdata = getpwnam(username);
+
+	/* Get password from user */
+	ttyattr.c_lflag &= ~(ECHO);
+	tcsetattr(STDIN_FILENO, TCSAFLUSH, &ttyattr);
+	psh_logincredentget(passwd, 1, maxlen);
+
+	/* validate against /etc/passwd */
+	if (userdata != NULL) {
+		shadow = crypt(passwd, userdata->pw_passwd);
+		if (strcmp(userdata->pw_passwd, shadow) == 0) {
+			auth = 1;
+		}
+	}
+
+#ifdef PSH_DEFUSRPASSWD
+	/* validate against defuser */
+	shadow = crypt(passwd, PSH_DEFUSRPASSWD);
+	if (auth != 1 && strcmp(username, "defuser") == 0 && strcmp(shadow, PSH_DEFUSRPASSWD) == 0) {
+			tcsetattr(STDIN_FILENO, TCSAFLUSH, &origattr);
+			auth = 1;
+	}
+#endif
+
+	ttyattr.c_lflag |= (ECHO);
+	tcsetattr(STDIN_FILENO, TCSAFLUSH, &ttyattr);
+	fprintf(stdout, "\n");
+	if (auth <= 0) {
+		sleep(2);
+		fprintf(stdout, "Login failed!\n");
+	}
+
+	tcsetattr(STDIN_FILENO, TCSAFLUSH, &origattr);
+	return auth;
```

Comment:
You should probably reset `passwd`.

Reply:
Password resetting added. 

After removing `auth` variable and adding multiple returns password resetting also has to be done multiple times (one for each return) which adds some code lines, but it is that or `auth` variable handling. Either way adds some lines.